doc/guide.md

```diff
@@ -0,0 +1,241 @@
+}
+```
+
+### Commands
+
+Each supported command must have its own function.
+
+The signature is:
+
+```c
+returncode_t command(<arguments>);
+```
+
+For every command, the function names must start with `[name]_command_` followed
+by a description of what the command is used for.
+
+### Example:
+
+```c
+returncode_t [name]_command_[desc](void) {
+  syscall_return_t cval = command(DRIVER_NUM_[NAME], 1, 0, 0);
+  return tock_command_return_novalue_to_returncode(cval);
+}
+```
+
+
+### Exists
+
+There must be a function to check for syscall driver existence.
+
+Signature:
+
+```
+bool [name]_exists(void);
+```
+
+Example:
+
+```c
+bool [name]_exists(void) {
+  return driver_exists(DRIVER_NUM_[NAME]);
+}
+```
+
+
+## Asynchronous APIs
+
+All common system call operations should have asynchronous versions.
+
+| Characteristic   | Value                         |
+|------------------|-------------------------------|
+| Location         | `libtock/[category]`          |
+| Source File Name | `libtock/[category]/[name].c` |
+| Header File Name | `libtock/[category]/[name].h` |
+
+For example, a sensor read operation should look like this:
+
+Define a suitable callback for the library:
+
+```c
+typedef void (*sensor_callback)(returncode_t, int);
+```
+
+Define a upcall function to be passed to the kernel:
+
+```c
+static void sensor_temp_upcall(int                          ret,
+                               int                          val,
+                               __attribute__ ((unused)) int unused0,
+                               void*                        ud) {
+  sensor_callback cb = (sensor_callback) ud;
+  cb(ret, val);
+}
+```
+
+Define an external function for applications to use to run the asynchronous
+operation:
+
+```c
+int sensor_read(sensor_callback cb) {
+  int ret;
+
+  ret = sensor_set_upcall(sensor_temp_upcall, cb);
+  if (ret != RETURNCODE_SUCCESS) return ret;
+
+  ret = sensor_command_read();
+  return ret;
+}
+```
+
+## Synchronous APIs
+
+Most system call interfaces will want to provide a synchronous API as well.
+
+| Characteristic   | Value                              |
+|------------------|------------------------------------|
+| Location         | `libtock-sync/[category]`          |
+| Source File Name | `libtock-sync/[category]/[name].c` |
+| Header File Name | `libtock-sync/[category]/[name].h` |
+
+For our sensor example:
+
+Create a struct to be set by the asynchronous callback:
+
+```c
+struct data {
+  bool fired;
+  int val;
+  returncode_t result;
+};
+
+static struct data result = { .fired = false };
+```
+
+Define a static callback function to pass to the asynchronous implementation:
+
+```c
+static void sensor_cb(returncode_t ret, int val) {
+  result.temp = val;
+  result.fired = true;
+  result.result = ret;
+}
+```
+
+Define the external function in the `libtocks_` name space for the sync
+operation:
+
+```c
+int libtocks_sensor_read(int* val) {
```

Comment:
Is `libtocks_` the proposed convention for synchronous functions?

Reply:
Yeah, I don't love it though.

Reply:
Even `libtocksync` would be better